src/transform/src/canonicalize_mfp.rs

```diff
@@ -129,4 +172,272 @@ impl CanonicalizeMfp {
+            *mfp = MapFilterProject::new(mfp.input_arity);
+            *mfp = mfp.clone().map(map);
+            *mfp = mfp.clone().filter(std::iter::once(new_pred));
+            *mfp = mfp.clone().project(project);
+
+            true
+        } else {
+            false
+        };
+    }
+
+    /// Makes the job of `literal_constraints` easier by undoing some CSE to reconstruct
+    /// literal constraints.
+    fn inline_literal_constraints(mfp: &mut MapFilterProject) -> Result<(), RecursionLimitError> {
+        let mut should_inline = vec![false; mfp.input_arity + mfp.expressions.len()];
+        // Mark those expressions for inlining that are of the form
+        // `<xxx> = <lit>` or `<lit> = <xxx>`.
+        for (i, e) in mfp.expressions.iter().enumerate() {
+            if let MirScalarExpr::CallBinary {
+                func: BinaryFunc::Eq,
+                expr1,
+                expr2,
+            } = e
+            {
+                if matches!(**expr1, MirScalarExpr::Literal(..))
+                    || matches!(**expr2, MirScalarExpr::Literal(..))
+                {
+                    should_inline[i + mfp.input_arity] = true;
+                }
+            }
+        }
+        // Whenever
+        // `<Column(i)> = <lit>` or `<lit> = <Column(i)>`
+        // appears in a predicate, mark the ith expression to be inlined.
+        for (_before, p) in mfp.predicates.iter() {
+            p.visit_post(&mut |e| {
+                if let MirScalarExpr::CallBinary {
+                    func: BinaryFunc::Eq,
+                    expr1,
+                    expr2,
+                } = e
+                {
+                    if matches!(**expr1, MirScalarExpr::Literal(..)) {
+                        if let MirScalarExpr::Column(col) = **expr2 {
+                            if col >= mfp.input_arity {
+                                should_inline[col] = true;
+                            }
+                        }
+                    }
+                    if matches!(**expr2, MirScalarExpr::Literal(..)) {
+                        if let MirScalarExpr::Column(col) = **expr1 {
+                            if col >= mfp.input_arity {
+                                should_inline[col] = true;
+                            }
+                        }
+                    }
+                }
+            })?;
+        }
+        // Perform the marked inlinings.
+        mfp.perform_inlining(should_inline);
+        Ok(())
+    }
+
+    /// MFPs have a Vec of predicates [p1, p2, ...], which logically represents `p1 AND p2 AND ...`.
+    /// This function performs this conversion. Note that it might create a variadic AND with
+    /// 0 or 1 args, so the resulting predicate Vec always has exactly 1 element.
+    fn list_of_predicates_to_and_of_predicates(mfp: &mut MapFilterProject) {
+        mfp.predicates = vec![(
+            mfp.input_arity + mfp.expressions.len(),
```

Comment:
I am in favor of rebuilding the MapFilterProject that way the `before` gets set to its minimal size.

Reply:
Done